### PR TITLE
Install a REJECT rule for nodeport with no backend

### LINKED ISSUE #43212: CLOSE_WAIT connections on master node when ELBs point there
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

CLOSE_WAIT

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.4", GitCommit:"7243c69eb523aa4377bce883e7c0dd76b84709a1", GitTreeState:"clean", BuildDate:"2017-03-08T02:48:58Z", GoVersion:"go1.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): Debian GNU/Linux 8 (jessie)
- **Kernel** (e.g. `uname -a`): Linux ip-172-31-64-14 4.4.41-k8s #1 SMP Mon Jan 9 15:34:39 UTC 2017 x86_64 GNU/Linux
- **Install tools**: kops 1.5.3
- **Others**:

**What happened**:

The problem was triggered when an Service type=LoadBalancer was left without ready Pods. This triggers a wave of CLOSE_WAIT on the master node(s) that is reproducible.

**What you expected to happen**:

There should not be any flooding of CLOSE_WAIT connections.

**How to reproduce it** (as minimally and precisely as possible):

- Start a cluster with Kops v1.5.3 (kubernetes v1.5.2) at AWS
- Create a Service type=LoadBalancer (without attached Pods)

This should trigger the CLOSE_WAIT on the master.

Take note that  because the kops v1.5.3 is using taints instead of SchedulingDisabled (https://github.com/kubernetes/kops/issues/639) the master nodes are also added under the ELB on AWS.

**Anything else we need to know**:

* As a workaround the master can be tagged as unscheduled. This will configure ELB to exclude the master and the CLOSE_WAITs will stop raising.

```
kubectl patch node MASTER_NAME -p "{\"spec\":{\"unschedulable\":true}}"
```
* If Pods are added to the LoadBalancer service, the CLOSE_WAITS will stop raising when ready.

* The CLOSE_WAITs start raising on master when ELB is tagging the master node as "InService" not before that.

* Once too many CLOSE_WAITs are generated the following error appears, and master is marked as `not_ready` and ssh is unresponsive. Logs were gathered from "AWS > Instance Settings > Get System Log"

```
TCP: out of memory… consider tuning tcp_mem
```

* Issue has been reproduced in a different cluster and AWS account.

reported together with @mikim83

### PR DESCRIPTION
**Pull request by thockin**
Rather than actually accepting the connection, REJECT.  This will avoid
CLOSE_WAIT.

Fixes #43212

@justinsb @felipejfc @spiddy

### CODE REVIEW

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -1108,6 +1108,18 @@ func (proxier *Proxier) syncProxyRules() {
 				// Currently we only create it for loadbalancers (#33586).
 				writeLine(natRules, append(args, "-j", string(svcXlbChain))...)
 			}
+
+			// If the service has no endpoints then reject packets.
+			if len(proxier.endpointsMap[svcName]) >= 0 {
```
Comment (justinsb) [inline pkg/proxy/iptables/proxier.go]: Do you mean `== 0` ?

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -1108,6 +1108,18 @@ func (proxier *Proxier) syncProxyRules() {
...
+				writeLine(filterRules,
+					"-A", string(kubeServicesChain),
+					"-m", "comment", "--comment", fmt.Sprintf(`"%s has no endpoints"`, svcName.String()),
```
Comment (justinsb) [inline pkg/proxy/iptables/proxier.go]: The comment here is inconsistent with `>= 0`

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -1108,6 +1108,18 @@ func (proxier *Proxier) syncProxyRules() {
 				// Currently we only create it for loadbalancers (#33586).
 				writeLine(natRules, append(args, "-j", string(svcXlbChain))...)
 			}
+
```
Comment (justinsb) [inline pkg/proxy/iptables/proxier.go]: We could do the "no endpoints" check before writing the MARK/SVC or XLB rules.  But I presume we want to keep the rules less volatile?

File: pkg/proxy/iptables/proxier.go [hunk 4]
```diff
@@ -1108,6 +1108,18 @@ func (proxier *Proxier) syncProxyRules() {
...
+			// If the service has no endpoints then reject packets.
```
Comment (justinsb) [inline pkg/proxy/iptables/proxier.go]: It took me a while to figure out that this is the analog of the nodeport entry chain below (e.g. I was wondering why you were changing kubeServicesChain and not the NodePort chain).  Expand the comment to that effect may be helpful.

e.g. 
```
If the service has no endpoints then reject packets.
We add a rule to the service chain that is the equivalent of the "tail-call to nodeports chain",
but immediately rejects the packet.
```

**Review by justinsb — COMMENTED** [review]
Some nits, but I haven't figured out how >= 0 is correct yet (but e2e didn't fail catastrophically, so...)

File: pkg/proxy/iptables/proxier.go (hunk 3 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: I have an idea for future refactoring that will make this nice, so I made the smallest change I could.

File: pkg/proxy/iptables/proxier.go (hunk 4 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: will add words

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: dangit, vestigial from something else I tried.

File: pkg/proxy/iptables/proxier.go (hunk 2 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: hush, you!

**Comment by thockin** [comment]
new push is up

**Comment by justinsb** [comment]
Sorry ... I see a new commit, but I don't see anything different?

File: pkg/proxy/iptables/proxier.go (hunk 2 above)
Comment (justinsb) [inline pkg/proxy/iptables/proxier.go]: Just wasn't sure which one was wrong :-)

File: pkg/proxy/iptables/proxier.go (hunk 3 above)
Comment (justinsb) [inline pkg/proxy/iptables/proxier.go]: You're saying you have a cunning plan?  :+1:

**Comment by justinsb** [comment]
BTW any idea why all the e2e tests aren't failing?  Aren't we blocking all NodePorts?

**Comment by thockin** [comment]
I botched the amend.  Fixed.  

It didn't trigger the e2es because the nat PREROUTING table triggers first, so the original packet never hits the filter table.

**Comment by justinsb** [comment]
/lgtm

Nice - and the explanation for how the e2e tests are not blowing up makes sense now you've explained it!

**Comment by justinsb** [comment]
Going to propose this for cherry picking to 1.4 and 1.5.

**Comment by justinsb** [comment]
So I'm afraid I think this is still happening in 1.6.0.  I can scale down a deployment to size 0 and I see CLOSE_WAIT accumulating, and I don't see any `has no endpoints` rule, either nodeport or clusterip.

I _think_ that the problem is maybe that we don't set `activeNATChains`, but I'm trying to get a test to fail.

**Comment by justinsb** [comment]
Edit: It does have the drop rules, in iptables -t filter (not -t nat). But the CLOSE_WAIT sockets are still accumulating. Digging...

**Comment by justinsb** [comment]
Figured it out I think - put details into #43969

**Comment by ketkulka** [comment]
@thockin the reject rules should also be installed for externalIPs and ports in addition to service IPs and node port. I am hitting a case where reject with icmp rule get installed for destination=service ip but not for destination=externalIPs. And when no pods are present; kube-proxy accepts the connection coming via externalIPs.

**Comment by thockin** [comment]
Can you file a new bug with a full repro, please?

**Comment by ketkulka** [comment]
Thanks @thockin .
Opened #44516 to track it. I can give a shot for the patch but I must admit I have no way of verifying all cases.

**Pull request merged**